Add packages for the OCaml system manual.

dbuenzli opened the pull request:
/cc @damiendoligez and @gasche 

I'm adding these packages that simply install the OCaml manual in `$(opam config var ocaml-manual:doc)`. The idea is that when installed, `opkg`'s documentation main index page will link directly on those according to your ocaml version rather than link to the latest version online.

I was tempted to patch the css file to "modernize" it a bit but didn't want to get into trouble with whoever is responsible for this. If there's interest in refreshing the look of the html manual upstream, I could give it a shot at some point.

avsm commented:
I like the packages, but would prefer for the ML files to be maintained out of opam-repository and in a separate upstream archive. Would setting something up on ocaml/ocaml-manual be appropriate @gasche ?

camelus commented:
##### :white_check_mark: All lint checks passed <small>ebb582100a483bdb0929aa82e3fba50904cf0d12</small>

* These packages passed lint tests: `ocaml-manual.4.01.0`, `ocaml-manual.4.02.0`, `ocaml-manual.4.03.0`


---


##### :white_check_mark: Installability check (5468 &rarr; 5471)

* new installable packages (3): `ocaml-manual.4.01.0` `ocaml-manual.4.02.0` `ocaml-manual.4.03.0`

dbuenzli commented:
@avsm that will be a little too complex to setup for what it is. Frankly the size of this file is not much larger than a `descr` with a long description. Can't simply next versions of the documentation tarball include that `install.ml` file ?

dbuenzli commented:
(Besides in my opinion the ocaml distribution should rather include the docs in the tarball and the ocaml opam package should install them in the `doc` directory of ocaml).

gasche commented:
It seems fine to me, and I including the script in future tarballs would be reasonable as far as I am concerned.

> I was tempted to patch the css file to "modernize" it a bit but didn't want to get into trouble with whoever is responsible for this. If there's interest in refreshing the look of the html manual upstream, I could give it a shot at some point.

I think that yes, there is interest (this has been done from time to time in the past), and that @damiendoligez and @Octachron would probably be the right person to discussi it with -- I suspect none of them are particularly likely to give you trouble.

dbuenzli commented:
> It seems fine to me, and I including the script in future tarballs would be reasonable as far as I am concerned.

Cool. That said this script is only here for windows compat I would happily have `cp`'d over to `ocaml-manual:doc` in the `build:` instructions otherwise.

dbuenzli commented:
In fact maybe if @dra27 gives us a command line and a filter for Windows we can skip remove the script and everybody will be happier.

Octachron commented:
> >I was tempted to patch the css file to "modernize" it a bit but didn't want to get into trouble with whoever is responsible for this. If there's interest in refreshing the look of the html manual upstream, I could give it a shot at some point.

>I think that yes, there is interest (this has been done from time to time in the past), and that @damiendoligez and @Octachron would probably be the right person to discussi it with -- I suspect none of them are particularly likely to give you trouble.

Personally, I think that refreshing the style of the manual would be a great idea. However, the last time that I played with the manual style, my conclusion was that the html layout itself would also probably need some improvements (for instance, to separate inputs and outputs within the code example environment).

dra27 commented:
@dbuenzli - as an aside, I set-up https://github.com/metastack/ocaml-legacy solely so that the OCaml packages can reference required Windows patches (using commit SHA), rather than embedding them in `files/`.

Currently, `cp` is an available command so you could just use it. The second phase of OPAM 2.0+Windows will look at hardening imported Cygwin commands, but this package is far from the only one using `cp` directly - `cp`, `mv`, and `rm` may end up as "default" imported commands anyway (along with `make`, `i686-mingw64-w32-gcc`, etc.).

To answer your question, though: filter is `{os = "win32"}` (doesn't capture Cygwin compilations of OPAM, which is correct) and the equivalent command is `"robocopy" "/E" "." ocaml-manual:doc`

dbuenzli commented:
Thanks @dra27 ! Removed the `install.ml` files, replaced by a multiplatform recursive copy in the `build:` section. @gasche nothing needs to be done, except it would be nice if in addition to the compiler package this could be added aswell on releases.

dsheets commented on packages/ocaml-manual/ocaml-manual.4.01.0/opam:
```diff
@@ -0,0 +1,19 @@
+available: [ ocaml-version >= "4.01.0" | ocaml-version < "4.02.0" ]
+build: [
+ [ "cp" "-R" "." ocaml-manual:doc "ocaml" {os != "win32"} ]
+ [ "robocopy" "/E" "." ocaml-manual:doc {os = "win32"} ]
```
These conditionals need to be for the line rather than the last component in the line (other side of `]`).

dbuenzli replied:
Uh... Thanks !

dbuenzli commented:
I think this can be merged.
